File: app/src/main/java/com/loftschool/kotlinexample/MainActivity.java
package com.loftschool.kotlinexample;

import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.Toast;

// https://antonioleiva.com/kotlin-awesome-tricks-for-android/
public class MainActivity extends AppCompatActivity {

    private LinearLayout container;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        container = (LinearLayout) findViewById(R.id.container);

        findViewById(R.id.add_item).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onAddItemClicked();
            }
        });

        findViewById(R.id.start_activity).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final Intent intent = new Intent(MainActivity.this, AnotherActivity.class);
                intent.putExtra(AnotherActivity.PARAM_NAME, "Some name");
                intent.putExtra(AnotherActivity.PARAM_PRICE, 15);
                startActivity(intent);
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_item:
                onAddItemClicked();
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private void onAddItemClicked() {
        addItem(new Item("item", 0));
        Snackbar.make(container, R.string.added, Snackbar.LENGTH_LONG)
                .setAction(R.string.cancel, new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Toast.makeText(MainActivity.this, R.string.cancelled, Toast.LENGTH_SHORT).show();
                    }
                })
                .setActionTextColor(Color.YELLOW)
                .show();
    }

    /// Items

    void addItem(Item item) {
        // adding item...
    }

    static class Item {
        String name;
        int price;

        Item(String name, int price) {
            this.name = name;
            this.price = price;
        }
    }

}

File: app/src/main/java/com/loftschool/kotlinexample/MainActivity.kt
package com.loftschool.kotlinexample

import android.graphics.Color
import android.os.Bundle
import android.support.annotation.ColorInt
import android.support.annotation.MenuRes
import android.support.annotation.StringRes
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import kotlinx.android.synthetic.main.activity_main.*
import org.jetbrains.anko.startActivity
import org.jetbrains.anko.toast

/**
 * Copyright (c) 2017.
 * All Rights Reserved.
 *
 * Author: Dmitry Gordeev <netimen@gmail.com>
 * Date:   24.04.17
 */

// https://antonioleiva.com/kotlin-awesome-tricks-for-android/
class MainActivityK : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        add_item.setOnClickListener { onAddItemClicked() }

        start_activity.setOnClickListener { startActivity<AnotherActivity>(AnotherActivity.PARAM_NAME to "Some name", AnotherActivity.PARAM_PRICE to 15) }
    }

//    override fun onCreateOptionsMenu(menu: Menu?): Boolean = inflateMenu(R.menu.main, menu)
    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        R.id.menu_item -> {
            onAddItemClicked()
            true
        }
//        R.id.menu_item -> consume { onAddItemClicked() }
        else -> super.onOptionsItemSelected(item)
    }

    fun onAddItemClicked() {
        addItem(Item("item", 0))
        Snackbar.make(container, R.string.added, Snackbar.LENGTH_LONG)
                .setAction(R.string.cancel) {
                    toast(R.string.cancelled)
                }
                .setActionTextColor(Color.YELLOW)
                .show()
//        container.snack(R.string.added) {
//            action(R.string.cancel, Color.YELLOW) { toast(R.string.cancelled) }
//        }
    }

    fun addItem(item: Item) {
        // adding item...
    }

}

class Item(val name: String, val price: Int)

/// SHOULD BE SOMEWHERE IN UTILS.KT

fun AppCompatActivity.inflateMenu(@MenuRes menuRes: Int, menu: Menu?): Boolean {
    menuInflater.inflate(menuRes, menu)
    return true
}

inline fun consume(f: () -> Unit): Boolean {
    f()
    return true
}

inline fun View.snack(@StringRes messageResId: Int, length: Int = Snackbar.LENGTH_LONG, f: Snackbar.() -> Unit) {
    val snack = Snackbar.make(this, messageResId, length)
    snack.f()
    snack.show()
}

fun Snackbar.action(@StringRes textResId: Int, @ColorInt color: Int? = null, listener: (View) -> Unit) {
    setAction(textResId, listener)
    color?.let { setActionTextColor(color) }
}
